gitlab/resource_gitlab_project.go

```diff
@@ -915,129 +915,142 @@ func resourceGitlabProjectDelete(ctx context.Context, d *schema.ResourceData, me
 func editOrAddPushRules(ctx context.Context, client *gitlab.Client, projectID string, d *schema.ResourceData) error {
 	log.Printf("[DEBUG] Editing push rules for project %q", projectID)
 
-	editOptions := expandEditProjectPushRuleOptions(d)
-	_, _, err := client.Projects.EditProjectPushRule(projectID, editOptions, gitlab.WithContext(ctx))
-	if err == nil {
-		return nil
-	}
+	pushRules, _, err := client.Projects.GetProjectPushRules(d.Id(), gitlab.WithContext(ctx))
+	// NOTE: push rules id `0` indicates that there haven't been any push rules set.
+	if err != nil || is404(err) || pushRules.ID == 0 {
+		if addOptions, needToAdd := expandAddProjectPushRuleOptions(d); needToAdd {
+			log.Printf("[DEBUG] Creating new push rules for project %q", projectID)
+			_, _, err = client.Projects.AddProjectPushRule(projectID, addOptions, gitlab.WithContext(ctx))
+			if err != nil {
+				return err
+			}
+		} else {
+			log.Printf("[DEBUG] Don't create new push rules for defaults for project %q", projectID)
+		}
 
-	var httpErr *gitlab.ErrorResponse
-	if !errors.As(err, &httpErr) || httpErr.Response.StatusCode != http.StatusNotFound {
-		return err
+		return nil
 	}
 
-	// A 404 could mean that the push rules need to be re-created.
-
-	log.Printf("[DEBUG] Failed to edit push rules for project %q: %v", projectID, err)
-	log.Printf("[DEBUG] Creating new push rules for project %q", projectID)
-
-	addOptions := expandAddProjectPushRuleOptions(d)
-	_, _, err = client.Projects.AddProjectPushRule(projectID, addOptions, gitlab.WithContext(ctx))
+	editOptions := expandEditProjectPushRuleOptions(d, pushRules)
+	log.Printf("[DEBUG] Editing existing push rules for project %q", projectID)
+	_, _, err = client.Projects.EditProjectPushRule(projectID, editOptions, gitlab.WithContext(ctx))
 	if err != nil {
 		return err
 	}
 
 	return nil
 }
 
-func expandEditProjectPushRuleOptions(d *schema.ResourceData) *gitlab.EditProjectPushRuleOptions {
+func expandEditProjectPushRuleOptions(d *schema.ResourceData, currentPushRules *gitlab.ProjectPushRules) *gitlab.EditProjectPushRuleOptions {
 	options := &gitlab.EditProjectPushRuleOptions{}
 
-	if d.HasChange("push_rules.0.author_email_regex") {
+	if d.HasChange("push_rules.0.author_email_regex") || d.Get("push_rules.0.author_email_regex") != currentPushRules.AuthorEmailRegex {
```

Comment:
On these `||` conditions, could you remove the left side and just check `d.Get("push_rules.0.author_email_regex") != currentPushRules.AuthorEmailRegex`?

Reply:
Also, does your test still pass if you removed all of the changes you made in this function? In other words is this needed at all?

Reply:
Yes, the pass at the moment even without this change, but ONLY if the instance defaults are the same as the provider defaults. See my comments here: https://github.com/gitlabhq/terraform-provider-gitlab/pull/838#discussion_r800220299

As soon as we can manage instance-wide push rules in the provider, we can test this .. may with some effort we could make that API call manually in the tests ...

Reply:
Sounds good. Since we only call this method after a call to `d.GetOk("push_rules.0")`, that should be ok. Can you hold off merging this until after we can release https://github.com/gitlabhq/terraform-provider-gitlab/milestone/5 ?

Reply:
Yes, I'll wait for that release for sure!